src/edu/csus/ecs/pc2/clics/API202306/CLICSContestAccess.java

```diff
@@ -0,0 +1,167 @@
+
+    /**
+     * Fill in properties for contest state as per 2023-06 spec
+     *
+     * @param sc security info for user making request
+     * @param model
+     * @param controller
+     * @param contestId
+     */
+    public CLICSContestAccess(SecurityContext sc, IInternalContest model, IInternalController controller, String contestId) {
+        this.controller = controller;
+
+        // For each role the connected user has, we enumerate what they can do with each endpoint.
+        ArrayList<String> cap = new ArrayList<String>();
+
+        if(ContestService.isContestStartAllowed(sc)) {
+            cap.add(API_CAPABILITY_CONTEST_START);
+        }
+        if(ContestService.isContestThawAllowed(sc)) {
+            cap.add(API_CAPABILITY_CONTEST_THAW);
+        }
+        if(SubmissionService.isTeamSubmitAllowed(sc)) {
+            cap.add(API_CAPABILITY_TEAM_SUBMIT);
+        }
+        if(ClarificationService.isTeamSubmitClarificationAllowed(sc)) {
+            cap.add(API_CAPABILITY_TEAM_CLAR);
+        }
+        if(SubmissionService.isProxySubmitAllowed(sc)) {
+            cap.add(API_CAPABILITY_PROXY_SUBMIT);
+        }
+        if(ClarificationService.isProxySubmitClarificationAllowed(sc)) {
+            cap.add(API_CAPABILITY_PROXY_CLAR);
+        }
+        if(SubmissionService.isAdminSubmitAllowed(sc)) {
+            cap.add(API_CAPABILITY_ADMIN_SUBMIT);
+        }
+        if(ClarificationService.isAdminSubmitClarificationAllowed(sc)) {
+            cap.add(API_CAPABILITY_ADMIN_CLAR);
+        }
+        if(AwardService.isAwardsUpdateAllowed(sc)) {
+            cap.add(API_CAPABILITY_AWARDS_UPDATE);
+        }
+        capabilities = cap.toArray(new String[0]);
+
+        /*
+         * As per the spec, here are the possible endpoints.
+         * These should match the class names in this package, eg. Contest -> ContestService, Run -> RunService, etc.
+         */
+        String [] serviceNames = { "Contest", "JudgementType", "Language", "Problem", "Group", "Organization",
+                "Team", "Person", "Account", "State", "Submission", "Judgement", "Run", "Clarification",
+                "Scoreboard", "EventFeed", "Award", "Commentary"
+        };
+
+        ArrayList<CLICSEndpoint> epList = new ArrayList<CLICSEndpoint>();
+
+        for (String service: serviceNames) {
+            if(!invokeServiceEndpointAccessMethod(epList, service, sc)) {
+                if(Utilities.isDebugMode()) {
+                    System.err.println("No endpoint access for service " + service);
+                }
+                controller.getLog().log(Log.WARNING, "No endpoint access for service " + service);
+            }
+        }
+        if(!epList.isEmpty()) {
+            endpoints = epList.toArray(new CLICSEndpoint[0]);
+        }
+    }
+
+    /**
+     * Create a fully qualified class name for the supplied web service endpoint.
+     * eg. Group would create: edu.csus.ecs.pc2clics.API202306.GroupService
```

Comment:
The "fully qualified class name" example above is missing a period (between "pc2" and "clics").

Reply:
Fixed in next push.